File: models/post.py
#!/usr/bin/python3
"""Defines the Post class"""
from sqlalchemy import Column, String, Text, ForeignKey
from sqlalchemy.orm import relationship
from models.base_model import BaseModel, Base
from models.like import Like

from models.post_tag import post_tag

class Post(BaseModel, Base):
    """Represents a blog post"""
    __tablename__ = 'posts'

    title = Column(String(255), nullable=False)
    author_id = Column(String(60), ForeignKey('users.id'), nullable=False)
    category_id = Column(String(60), ForeignKey('categories.id'), nullable=True)

    # Relationships
    category = relationship('Category', back_populates='posts')
    contents = relationship("Content", back_populates="post", cascade="all, delete-orphan")
    author = relationship("User", back_populates="posts")
    comments = relationship("Comment", back_populates="post", cascade="all, delete-orphan")
    likes = relationship("Like", back_populates="post", cascade="all, delete-orphan")
    tags = relationship("Tag", secondary=post_tag, back_populates="posts")

    def count_comment(self):
        """Return the number of comment for the post."""
        return len(self.comments)

    def count_likes(self):
        """Return the number of likes for the post."""
        return len(self.likes)

    def is_liked(self, user, post):
        from models import storage
        likes = storage.all(Like).values()
        for like in likes:
            if like.post_id == post:
                if like.author_id == user:
                    return True
        return False

    def add_tag(self, tag):
        """Add a tag to the post."""
        if tag not in self.tags:
            self.tags.append(tag)
            models.storage.save()

    def remove_tag(self, tag):
        """Remove a tag from the post."""
        if tag in self.tags:
            self.tags.remove(tag)
            models.storage.save()

File: api/models/post.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// Post represents a blog post.
type Post struct {
	ID        uuid.UUID `gorm:"type:char(36);primaryKey" json:"id"`
	Title     string    `gorm:"type:varchar(255);not null" json:"title"`
	AuthorID  uuid.UUID `gorm:"type:char(36);not null" json:"author_id"`
	CreatedAt time.Time `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt time.Time `gorm:"autoUpdateTime" json:"updated_at"`

	// Relationships
	Reports  []Report  `gorm:"foreignKey:PostID;constraint:OnDelete:SET NULL" json:"reports"`
	Contents []Content `gorm:"foreignKey:PostID;constraint:OnDelete:CASCADE" json:"contents"`
	Comments []Comment `gorm:"foreignKey:PostID;constraint:OnDelete:SET NULL" json:"comments"`
	Likes    []Like    `gorm:"foreignKey:PostID;constraint:OnDelete:CASCADE" json:"likes"`
	Tags     []Tag     `gorm:"many2many:post_tags;" json:"tags"`

	Author User `gorm:"foreignKey:AuthorID" json:"author"`
}

// CountComments returns the number of comments for the post.
func (p *Post) CountComments() int {
	return len(p.Comments)
}

// CountLikes returns the number of likes for the post.
func (p *Post) CountLikes() int {
	return len(p.Likes)
}

// IsLiked checks whether a user has liked the post.
func (p *Post) is_liked(userID uuid.UUID, db *gorm.DB) bool {
	var like Like
	result := db.Where("post_id = ? AND author_id = ?", p.ID, userID).First(&like)
	return result.RowsAffected > 0
}

// AddTag adds a tag to the post.
func (p *Post) AddTag(tag Tag, db *gorm.DB) error {
	err := db.Model(&p).Association("Tags").Append(&tag)
	return err
}

// RemoveTag removes a tag from the post.
func (p *Post) RemoveTag(tag Tag, db *gorm.DB) error {
	err := db.Model(&p).Association("Tags").Delete(&tag)
	return err
}

// BeforeCreate hook ensures UUID is set for ID field
func (p *Post) BeforeCreate(tx *gorm.DB) (err error) {
	if p.ID == uuid.Nil {
		p.ID = uuid.New() // Generate a new UUID for ID if not set
	}
	return
}
